src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm

```diff
@@ -1246,6 +1260,42 @@ - (PersistentStorageDelegate *)storageDelegate
     return &_groupDataProvider;
 }
 
+#pragma mark - Controller suspension support
+
+- (BOOL)_allControllersSuspended
+{
+    auto * controllers = [self getRunningControllers];
+    for (MTRDeviceController * controller in controllers) {
+        if (!controller.suspended) {
+            return NO;
+        }
+    }
+    return YES;
+}
+
+- (void)controller:(MTRDeviceController *)controller
+       isSuspended:(BOOL)suspended
+{
+    // We use the Matter queue as our delegate queue for this notification.
+    assertChipStackLockedByCurrentThread();
+
+    if (!suspended) {
+        // As long as we have one controller that is not suspended, we should
+        // have a running operational browse. Guard against someone resuming a
```

Comment:
Shouldn't resuming a controller that's shut down fail to begin with, or at least not emit events to the delegate?

Reply:
Ideally yes, but right now it does not.  Once we fix it to do that, we can remove this part.